// @aws-cdk/cloudformation-diff/lib/diffable.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DiffableCollection = void 0;
/**
 * Calculate differences of immutable elements
 */
class DiffableCollection {
    constructor() {
        this.additions = [];
        this.removals = [];
        this.oldElements = [];
        this.newElements = [];
    }
    addOld(...elements) {
        this.oldElements.push(...elements);
    }
    addNew(...elements) {
        this.newElements.push(...elements);
    }
    calculateDiff() {
        this.additions.push(...difference(this.newElements, this.oldElements));
        this.removals.push(...difference(this.oldElements, this.newElements));
    }
    get hasChanges() {
        return this.additions.length + this.removals.length > 0;
    }
    get hasAdditions() {
        return this.additions.length > 0;
    }
    get hasRemovals() {
        return this.removals.length > 0;
    }
}
exports.DiffableCollection = DiffableCollection;
/**
 * Whether a collection contains some element (by value)
 */
function contains(element, xs) {
    return xs.some(x => x.equal(element));
}
/**
 * Return collection except for elements
 */
function difference(collection, elements) {
    return collection.filter(x => !contains(x, elements));
}
