// aws-cdk/test/bockfs.js
"use strict";
// A not-so-fake filesystem mock similar to mock-fs
//
// mock-fs is super convenient but we can't always use it:
// - When you use console.log() jest wants to load things from the filesystem (which fails).
// - When you make AWS calls the SDK wants to load things from the filesystem (which fails).
//
// Therefore, something similar which uses tempdirs on your actual disk.
//
// The big downside compared to mockfs is that you need to use bockfs.path() to translate
// fake paths to real paths.
const os = require("os");
const path_ = require("path");
const fs = require("fs-extra");
const bockFsRoot = fs.mkdtempSync(path_.join(os.tmpdir(), 'bockfs'));
let oldCwd;
function bockfs(files) {
    oldCwd = process.cwd();
    for (const [fileName, contents] of Object.entries(files)) {
        bockfs.write(fileName, contents);
    }
}
(function (bockfs) {
    /**
     * Write contents to a fake file
     */
    function write(fakeFilename, contents) {
        const fullPath = path(fakeFilename);
        fs.mkdirSync(path_.dirname(fullPath), { recursive: true });
        fs.writeFileSync(fullPath, contents, { encoding: 'utf-8' });
    }
    bockfs.write = write;
    /**
     * Turn a fake path into a real path
     */
    function path(fakePath) {
        if (fakePath.startsWith('/')) {
            fakePath = fakePath.substr(1);
        } // Force path to be non-absolute
        return path_.join(bockFsRoot, fakePath);
    }
    bockfs.path = path;
    /**
     * Change to a fake directory
     */
    function workingDirectory(fakePath) {
        process.chdir(path(fakePath));
    }
    bockfs.workingDirectory = workingDirectory;
    function executable(...fakePaths) {
        for (const fakepath of fakePaths) {
            fs.chmodSync(path(fakepath), '755');
        }
    }
    bockfs.executable = executable;
    /**
     * Remove all files and restore working directory
     */
    function restore() {
        if (oldCwd) {
            process.chdir(oldCwd);
        }
        fs.removeSync(bockFsRoot);
    }
    bockfs.restore = restore;
})(bockfs || (bockfs = {}));
module.exports = bockfs;
