// @aws-cdk/aws-ec2/lib/ec2-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.VPNMetrics = exports.TransitGatewayMetrics = exports.NATGatewayMetrics = exports.EC2Metrics = exports.EBSMetrics = exports.EC2CapacityReservationsMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class EC2CapacityReservationsMetrics {
    static instanceUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/EC2CapacityReservations',
            metricName: 'InstanceUtilization',
            dimensions,
            statistic: 'Average',
        };
    }
    static usedInstanceCountAverage(dimensions) {
        return {
            namespace: 'AWS/EC2CapacityReservations',
            metricName: 'UsedInstanceCount',
            dimensions,
            statistic: 'Average',
        };
    }
    static availableInstanceCountAverage(dimensions) {
        return {
            namespace: 'AWS/EC2CapacityReservations',
            metricName: 'AvailableInstanceCount',
            dimensions,
            statistic: 'Average',
        };
    }
    static totalInstanceCountAverage(dimensions) {
        return {
            namespace: 'AWS/EC2CapacityReservations',
            metricName: 'TotalInstanceCount',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.EC2CapacityReservationsMetrics = EC2CapacityReservationsMetrics;
class EBSMetrics {
    static volumeReadBytesSum(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'VolumeReadBytes',
            dimensions,
            statistic: 'Sum',
        };
    }
    static volumeWriteBytesSum(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'VolumeWriteBytes',
            dimensions,
            statistic: 'Sum',
        };
    }
    static volumeReadOpsSum(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'VolumeReadOps',
            dimensions,
            statistic: 'Sum',
        };
    }
    static volumeTotalReadTimeAverage(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'VolumeTotalReadTime',
            dimensions,
            statistic: 'Average',
        };
    }
    static volumeWriteOpsSum(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'VolumeWriteOps',
            dimensions,
            statistic: 'Sum',
        };
    }
    static volumeTotalWriteTimeAverage(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'VolumeTotalWriteTime',
            dimensions,
            statistic: 'Average',
        };
    }
    static volumeIdleTimeAverage(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'VolumeIdleTime',
            dimensions,
            statistic: 'Average',
        };
    }
    static volumeQueueLengthAverage(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'VolumeQueueLength',
            dimensions,
            statistic: 'Average',
        };
    }
    static burstBalanceAverage(dimensions) {
        return {
            namespace: 'AWS/EBS',
            metricName: 'BurstBalance',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.EBSMetrics = EBSMetrics;
class EC2Metrics {
    static cpuUtilizationAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'CPUUtilization',
            dimensions,
            statistic: 'Average',
        };
    }
    static statusCheckFailedSum(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'StatusCheckFailed',
            dimensions,
            statistic: 'Sum',
        };
    }
    static statusCheckFailedInstanceSum(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'StatusCheckFailed_Instance',
            dimensions,
            statistic: 'Sum',
        };
    }
    static statusCheckFailedSystemSum(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'StatusCheckFailed_System',
            dimensions,
            statistic: 'Sum',
        };
    }
    static diskReadBytesAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'DiskReadBytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static diskReadOpsAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'DiskReadOps',
            dimensions,
            statistic: 'Average',
        };
    }
    static diskWriteBytesAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'DiskWriteBytes',
            dimensions,
            statistic: 'Average',
        };
    }
    static diskWriteOpsAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'DiskWriteOps',
            dimensions,
            statistic: 'Average',
        };
    }
    static networkInAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'NetworkIn',
            dimensions,
            statistic: 'Average',
        };
    }
    static networkOutAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'NetworkOut',
            dimensions,
            statistic: 'Average',
        };
    }
    static networkPacketsInAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'NetworkPacketsIn',
            dimensions,
            statistic: 'Average',
        };
    }
    static networkPacketsOutAverage(dimensions) {
        return {
            namespace: 'AWS/EC2',
            metricName: 'NetworkPacketsOut',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.EC2Metrics = EC2Metrics;
class NATGatewayMetrics {
    static activeConnectionCountMaximum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'ActiveConnectionCount',
            dimensions,
            statistic: 'Maximum',
        };
    }
    static packetsDropCountSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'PacketsDropCount',
            dimensions,
            statistic: 'Sum',
        };
    }
    static bytesInFromDestinationSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'BytesInFromDestination',
            dimensions,
            statistic: 'Sum',
        };
    }
    static bytesInFromSourceSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'BytesInFromSource',
            dimensions,
            statistic: 'Sum',
        };
    }
    static bytesOutToDestinationSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'BytesOutToDestination',
            dimensions,
            statistic: 'Sum',
        };
    }
    static bytesOutToSourceSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'BytesOutToSource',
            dimensions,
            statistic: 'Sum',
        };
    }
    static connectionAttemptCountSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'ConnectionAttemptCount',
            dimensions,
            statistic: 'Sum',
        };
    }
    static connectionEstablishedCountSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'ConnectionEstablishedCount',
            dimensions,
            statistic: 'Sum',
        };
    }
    static errorPortAllocationSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'ErrorPortAllocation',
            dimensions,
            statistic: 'Sum',
        };
    }
    static idleTimeoutCountSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'IdleTimeoutCount',
            dimensions,
            statistic: 'Sum',
        };
    }
    static packetsInFromDestinationSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'PacketsInFromDestination',
            dimensions,
            statistic: 'Sum',
        };
    }
    static packetsInFromSourceSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'PacketsInFromSource',
            dimensions,
            statistic: 'Sum',
        };
    }
    static packetsOutToDestinationSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'PacketsOutToDestination',
            dimensions,
            statistic: 'Sum',
        };
    }
    static packetsOutToSourceSum(dimensions) {
        return {
            namespace: 'AWS/NATGateway',
            metricName: 'PacketsOutToSource',
            dimensions,
            statistic: 'Sum',
        };
    }
}
exports.NATGatewayMetrics = NATGatewayMetrics;
class TransitGatewayMetrics {
    static bytesInSum(dimensions) {
        return {
            namespace: 'AWS/TransitGateway',
            metricName: 'BytesIn',
            dimensions,
            statistic: 'Sum',
        };
    }
    static bytesOutSum(dimensions) {
        return {
            namespace: 'AWS/TransitGateway',
            metricName: 'BytesOut',
            dimensions,
            statistic: 'Sum',
        };
    }
    static packetDropCountBlackholeSum(dimensions) {
        return {
            namespace: 'AWS/TransitGateway',
            metricName: 'PacketDropCountBlackhole',
            dimensions,
            statistic: 'Sum',
        };
    }
    static packetDropCountNoRouteSum(dimensions) {
        return {
            namespace: 'AWS/TransitGateway',
            metricName: 'PacketDropCountNoRoute',
            dimensions,
            statistic: 'Sum',
        };
    }
    static packetsInSum(dimensions) {
        return {
            namespace: 'AWS/TransitGateway',
            metricName: 'PacketsIn',
            dimensions,
            statistic: 'Sum',
        };
    }
    static packetsOutSum(dimensions) {
        return {
            namespace: 'AWS/TransitGateway',
            metricName: 'PacketsOut',
            dimensions,
            statistic: 'Sum',
        };
    }
}
exports.TransitGatewayMetrics = TransitGatewayMetrics;
class VPNMetrics {
    static tunnelDataInSum(dimensions) {
        return {
            namespace: 'AWS/VPN',
            metricName: 'TunnelDataIn',
            dimensions,
            statistic: 'Sum',
        };
    }
    static tunnelStateAverage(dimensions) {
        return {
            namespace: 'AWS/VPN',
            metricName: 'TunnelState',
            dimensions,
            statistic: 'Average',
        };
    }
    static tunnelDataOutSum(dimensions) {
        return {
            namespace: 'AWS/VPN',
            metricName: 'TunnelDataOut',
            dimensions,
            statistic: 'Sum',
        };
    }
}
exports.VPNMetrics = VPNMetrics;
