// @aws-cdk/aws-events/lib/events-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventsMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class EventsMetrics {
    static invocationsAverage(dimensions) {
        return {
            namespace: 'AWS/Events',
            metricName: 'Invocations',
            dimensions,
            statistic: 'Average',
        };
    }
    static failedInvocationsAverage(dimensions) {
        return {
            namespace: 'AWS/Events',
            metricName: 'FailedInvocations',
            dimensions,
            statistic: 'Average',
        };
    }
    static triggeredRulesAverage(dimensions) {
        return {
            namespace: 'AWS/Events',
            metricName: 'TriggeredRules',
            dimensions,
            statistic: 'Average',
        };
    }
    static matchedEventsAverage(dimensions) {
        return {
            namespace: 'AWS/Events',
            metricName: 'MatchedEvents',
            dimensions,
            statistic: 'Average',
        };
    }
    static deadLetterInvocationsAverage(dimensions) {
        return {
            namespace: 'AWS/Events',
            metricName: 'DeadLetterInvocations',
            dimensions,
            statistic: 'Average',
        };
    }
}
exports.EventsMetrics = EventsMetrics;
