// @aws-cdk/cfnspec/test/test.build.js
"use strict";
const massage_spec_1 = require("../build-tools/massage-spec");
module.exports = {
    'dropTypelessAttributes works correctly'(test) {
        const spec = {
            Fingerprint: 'some-fingerprint',
            PropertyTypes: {
                'CDK::Test::Property': {
                    Properties: {
                        Type: {
                            PrimitiveType: 'String',
                        },
                    },
                },
            },
            ResourceTypes: {
                'CDK::Test::Resource': {
                    Attributes: {
                        Attribute1: {
                            PrimitiveType: 'String',
                        },
                        Attribute2: {},
                    },
                    Documentation: 'https://documentation-url/cdk/test/resource',
                    Properties: {
                        ResourceArn: {
                            PrimitiveType: 'String',
                        },
                    },
                },
            },
        };
        massage_spec_1.massageSpec(spec);
        test.deepEqual(spec, {
            Fingerprint: 'some-fingerprint',
            PropertyTypes: {
                'CDK::Test::Property': {
                    Properties: {
                        Type: {
                            PrimitiveType: 'String',
                        },
                    },
                },
            },
            ResourceTypes: {
                'CDK::Test::Resource': {
                    Attributes: {
                        Attribute1: ({
                            PrimitiveType: 'String',
                        }),
                    },
                    Documentation: 'https://documentation-url/cdk/test/resource',
                    Properties: {
                        ResourceArn: {
                            PrimitiveType: 'String',
                        },
                    },
                },
            },
        });
        test.done();
    },
};
