// @aws-cdk/aws-logs/lib/logs-canned-metrics.generated.js
"use strict";
// Copyright 2012-2021 Amazon.com, Inc. or its affiliates. All Rights Reserved.
Object.defineProperty(exports, "__esModule", { value: true });
exports.LogsMetrics = void 0;
/* eslint-disable max-len */ // This is generated code - line lengths are difficult to control
class LogsMetrics {
    static incomingLogEventsSum(dimensions) {
        return {
            namespace: 'AWS/Logs',
            metricName: 'IncomingLogEvents',
            dimensions,
            statistic: 'Sum',
        };
    }
    static deliveryErrorsSum(dimensions) {
        return {
            namespace: 'AWS/Logs',
            metricName: 'DeliveryErrors',
            dimensions,
            statistic: 'Sum',
        };
    }
    static incomingBytesSum(dimensions) {
        return {
            namespace: 'AWS/Logs',
            metricName: 'IncomingBytes',
            dimensions,
            statistic: 'Sum',
        };
    }
    static forwardedBytesSum(dimensions) {
        return {
            namespace: 'AWS/Logs',
            metricName: 'ForwardedBytes',
            dimensions,
            statistic: 'Sum',
        };
    }
    static forwardedLogEventsSum(dimensions) {
        return {
            namespace: 'AWS/Logs',
            metricName: 'ForwardedLogEvents',
            dimensions,
            statistic: 'Sum',
        };
    }
    static deliveryThrottlingSum(dimensions) {
        return {
            namespace: 'AWS/Logs',
            metricName: 'DeliveryThrottling',
            dimensions,
            statistic: 'Sum',
        };
    }
}
exports.LogsMetrics = LogsMetrics;
